### PR TITLE
Feature/DocumentStore

### PR DESCRIPTION
**Pull request by vinodkiran**
Adding the Document Store functionality....

### CODE REVIEW

File: packages/ui/src/views/docstore/AddDocStoreDialog.jsx [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+                        )
+                    }
+                })
+                onConfirm(saveResp.data.id)
+            }
+        } catch (error) {
+            const errorData = error.response?.data || `${error.response?.status}: ${error.response?.statusText}`
+            enqueueSnackbar({
+                message: `Failed to update Document Store: ${errorData}`,
+                options: {
```
Comment (HenryHengZJ) [inline packages/ui/src/views/docstore/AddDocStoreDialog.jsx:136]: `message: `Failed to add new Document Store: ${error.response.data.message}`,`

File: packages/server/src/routes/index.ts [hunk 2]
```diff
@@ -48,6 +49,7 @@ router.use('/components-credentials', componentsCredentialsRouter)
 router.use('/components-credentials-icon', componentsCredentialsIconRouter)
 router.use('/chatflows-uploads', chatflowsUploadsRouter)
 router.use('/credentials', credentialsRouter)
+router.use('/documentStore', documentStoreRouter)
```
Comment (HenryHengZJ) [inline packages/server/src/routes/index.ts:52]: should we follow the naming convention `document-store`

File: packages/server/src/controllers/documentstore/index.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        if (typeof req.body === 'undefined') {
+            throw new Error(`Error: documentStoreController.createDocumentStore - body not provided!`)
+        }
+        const body = req.body
+        const subFolder = convertToValidFilename(body.name)
+        const dir = path.join(getStoragePath(), 'datasource', subFolder)
+        if (fs.existsSync(dir)) {
+            return res.status(500).send(new Error(`Document store ${body.name} already exists. Subfolder: ${subFolder}`))
+        } else {
+            fs.mkdirSync(dir)
```
Comment (HenryHengZJ) [inline packages/server/src/controllers/documentstore/index.ts:21]: `fs.mkdirSync(dir, { recursive: true })`

File: packages/server/src/dto/DocumentStoreDTO.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
+import { DocumentStoreStatus } from '../Interface'
```
Comment (HenryHengZJ) [inline packages/server/src/dto/DocumentStoreDTO.ts:1]: what does `dto` stands for?

File: packages/server/src/dto/DocumentStoreDTO.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            documentStoreDTO.files = JSON.parse(entity.files)
+        }
+        return documentStoreDTO
+    }
+
+    static fromEntities(entities: DocumentStore[]): DocumentStoreDTO[] {
+        return entities.map((entity) => this.fromEntity(entity))
+    }
+
+    static toEntity(body: any): DocumentStore {
```
Comment (HenryHengZJ) [inline packages/server/src/dto/DocumentStoreDTO.ts:55]: whenever we `updateDocumentStore`, this method is called, and everything is then reset like `files`, `whereUsed` etc

Comment (HenryHengZJ) [reply packages/server/src/dto/DocumentStoreDTO.ts:55]: I didnt see `whereUsed` being used anywhere in other places, whats the purpose of that?

File: packages/ui/src/views/docstore/AddDocStoreDialog.jsx (hunk 1 above)
Comment (vinodkiran) [reply packages/ui/src/views/docstore/AddDocStoreDialog.jsx:136]: fixed.

File: packages/server/src/routes/index.ts (hunk 2 above)
Comment (vinodkiran) [reply packages/server/src/routes/index.ts]: done

File: packages/server/src/dto/DocumentStoreDTO.ts (hunk 5 above)
Comment (vinodkiran) [reply packages/server/src/dto/DocumentStoreDTO.ts:55]: > whenever we `updateDocumentStore`, this method is called, and everything is then reset like `files`, `whereUsed` etc

Fixed.

Comment (vinodkiran) [reply packages/server/src/dto/DocumentStoreDTO.ts:55]: whereUsed is part of functionality in progress. Will hold an array of chatflow id that use this DocumentStore. Will be used to warn the user when they edit / add files / change chunking config on an existing doc store

File: packages/server/src/dto/DocumentStoreDTO.ts (hunk 4 above)
Comment (vinodkiran) [reply packages/server/src/dto/DocumentStoreDTO.ts:1]: [DTO Pattern](https://dev.to/tamj0rd2/dto-a-typescript-utility-type-4o3m).

Using a DTO as the entity as a number of stringified JSON objects, the front end will not need to see the underlying entity structure

File: packages/server/src/Interface.DocumentStore.ts [hunk 6]
```diff
@@ -1,22 +1,94 @@
-import { DocumentStoreStatus } from '../../Interface'
-import { DocumentStore } from '../../database/entities/DocumentStore'
+import { DocumentStore } from './database/entities/DocumentStore'
```
Comment (HenryHengZJ) [inline packages/server/src/Interface.DocumentStore.ts:1]: just so we are consistent, are we separating the interfaces like this for each entity going forward? Is this the best way?

Comment (vinodkiran) [reply packages/server/src/Interface.DocumentStore.ts:1]: With other large modules on the way, a single Interface.ts was becoming too large and unmanageable.

**Review by HenryHengZJ — APPROVED** [review]

**Comment by ill-yes** [comment]
Thank you very much! I've been waiting so long for this fix! 😭🙏🏽

**Pull request merged**